Декларировать БЭМ-сущности из всех ключей входящего bemjson-а

narqo opened the pull request:
Сейчас декарация входного BEMJSON строится только по предопределенным ключам (`block`, `elem`, `mod`, `content`, пр.)

Нужна возможность строить декларацию по всем ключам, для более гибкого описания входных данным:

```
block: 'blah',
meta: {
   elem: 'el'
},
content: {

}
```

Commit:
Декларировать БЭМ-сущности из всех ключей входящего bemjson-а

varya commented:
А насколько это преобразование дольше происходит?

veged commented on lib/techs/bemdecl.js.js:
```diff
@@ -72,7 +72,13 @@ function getBuilder(decl, block) {
 
-        iterateJson([obj.mix, obj.content], getBuilder(decl, block));
+        props = (Object.keys(obj)).filter(function(k) {
+                return ['block', 'elem', 'mod', 'elemMods'].indexOf(k) === -1;
```
`({ block: 1, elem: 1, mod: 1, elemMods: 1 }).hasOwnProperty(k)` быстрее чем итерация в `indexOf`

veged commented on lib/techs/bemdecl.js.js:
```diff
@@ -72,7 +72,13 @@ function getBuilder(decl, block) {
                     });
 
-        iterateJson([obj.mix, obj.content], getBuilder(decl, block));
+        props = (Object.keys(obj)).filter(function(k) {
```
этот `filter` вообще можно не делать, можно просто сразу `iterateJson(obj, ...`, на простых значениях из block, elem, mod и elemMods не наэкономишь -- у меня первоначально опасения были просто за левый контент кастомных данных, который в принципе в больших количествах может лежать в БЭМ-сущностях

Commit:
Faster object keys filter

arikon commented:
@veged, возьмёшь на себя этот PR?

Commit:
Boolean type added to isSimple

veged commented:
@arikon да

veged commented:
@narqo см. мой камент к L75

Commit:
Iterate through all nodes in bemjson

arikon commented:
@veged @narqo Будете продолжать?

veged commented:
да вроде можно мёржить, нужно только с HEAD актуализировать

arikon committed:
Add util.uniq() generator of callbacks for [].reduce()

arikon committed:
Add ability to customize build rules more flexibly by providing Arch.createCustomNode() method

Commit:
Tests bem-bl submodule is updated to 0.2 (close #264)

- tests bem-bl updated
- tests fixed to inherit i18n.js from js-i to build files with include directives (close #265)

arikon committed:
js+coffee and js+coffee-i tech modules

- Add `js+coffee` and `js+coffee-i` tech modules
- Refactor `js` and `js-i` tech modules
- BundleNode: Use `js+coffee` tech instead of `js` for `BorschikNode` of `js` techs by default
- package: Depends on `coffee-script` and `borschik ~0.0.13`
- tests: Update `bem-bl 0.2` (`i18n.js` tech module)
- tests: Use `js+coffee-i` as `js` tech to build bundles

arikon committed:
Some more tuning and jsdoc

arikon committed:
Fix link

Commit:
README is splat into several files

arikon committed:
Add level config with `md` tech for `docs/`

arikon committed:
Merge pull request #273 from bem/doc-split

`README` is splat into several files in `docs/`

banzalik committed:
Update README.ru.md

опечатка помошью -> помощью

vithar committed:
Merge pull request #275 from banzalik/patch-4

Update README.ru.md

arikon committed:
GitLibraryNode: Fix checkout of the last commit

There was an error checking out last commit of the branch:

- checkout of repos with only one commit didn't work
- checkout of branch was touching files from the last commit, it was affecting build processes, as these files was always updated

arikon committed:
Refactor of require() function in `.bem/make.js` files, it is now fully context aware

arikon committed:
Refactor of require() function in declaration files, it is now fully context aware

arikon committed:
util: Introduce `isFileValid()` and `itTimeValid()` convenient utility functions

The are used in `bem-www` `.bem/make.js`, for example

apsavin committed:
Docs fix: bemhtml.js -> bemhtml

arikon committed:
Merge pull request #276 from apsavin/docs-fix

Docs fix: bemhtml.js -> bemhtml

arikon committed:
Create node_modules/bem -> .. symlink on module install (Close #277)

arikon committed:
Do not make symlink on Travis build run

arikon committed:
Merge branch 'master' into bem-www

Commit:
Part of the default-arch is moved to standalone node (SeedNode) (close #87)
Libraries utilize the same build initialization as project - it's possible to build blocks levels within libraries (close #52)

arikon committed:
Add .editorconfig (Close #278)

arikon committed:
Add .editorconfig: order is necessary

arikon committed:
Refactor Tech: introduce getBuildSuffixesMap()

arikon committed:
Add match*() methods to `simple` level prototype, add tests (Close #282)

arikon committed:
Merge pull request #279 from bem/suffixes-map

Refactor Tech: introduce getBuildSuffixesMap()

arikon committed:
Merge branch 'master' into js+coffee

arikon committed:
Simplify coffee+js.js tech (Ref bem/bem-tools#279)

arikon committed:
Use latest bem-bl with updated i18n.js.js tech to reflect changes in coffee+js.js tech

arikon committed:
Use latest xjst and ometajs to fix tests

- Small cleanup of `js+coffee.js` code
- Use updated `bem-bl`
- Canonize new reference test results

arikon committed:
Merge pull request #268 from bem/js+coffee

js+coffee and js+coffee-i tech modules

arikon committed:
Fix recursion error when build target name contain trailing slash (Close #252)

arikon committed:
Merge pull request #272 from bem/bem-www

Arch: Add ability to customize build rules more flexibly

arikon committed:
Add `dir` tech module

arikon committed:
Merge pull request #290 from bem/dir-tech

Tech module to create directories

vithar committed:
Simple index page added

vithar committed:
markdown list markup fixed

arikon committed:
Replace own `util.extend()` implementation with `node.extend` module

tadatuta committed:
combine bem commands documentation

tadatuta committed:
Doc: Update index page to fit new doc structure

arikon committed:
Fix jsdoc in util

arikon committed:
Slightly refactor `bem create level` command

- Move validation of level prototype logic to the `COA.Opt.val()` method
- More promisy-based code
- Code cleanup
- Updated reference legacy-tests results

arikon committed:
Use promisy function to write level .js module

arikon committed:
Add ability to `bem create level` command to create level .js modules

arikon committed:
Use coa ~0.3.6

arikon committed:
Fix jsdoc of util.exec(): `options` and `resolveWithOutput` are optional arguments

arikon committed:
Fix jsdoc on util.relative(): mark `dot` argument as optional

arikon committed:
Add util.bemParseKey() function and tests

arikon committed:
Add `.bem` and `node_modules` to `Level.isIgnorablePath()`

arikon committed:
Fix jsdoc of `Tech.getPath()` and `Tech.getPaths()` methods: mark `suffix` and `suffixes` arguments as optional

arikon committed:
bem create: Ability to customize default techs list based on BEM-entity

Commit:
Inspector is improved

The graph is shown collapsed when snapshot is chosen. Click a node to expand/collapse it.

Commit:
MergedBundle improvements

MergedBundle allows to override the list of the bundles to use.
The tech to merge (bemdecl.js, deps.js) can be overridden.

Commit:
A typo in logging message

Commit:
Описал пример поключения локализационных технологий сборки

Commit:
Исправил замечания из #301

Commit:
Добавил английский вариант описания

Commit:
Исправил опечатку в описании

Commit:
Git library checkout fixed to work with commit hashes (close #302)

Parameter branch is added to specify branch name. Use treeish to specify
commit or tag.

tadatuta committed:
Doc: Fix techs link

Commit:
Magic nodes doesn't link the nodes it creates with parent magic nodes
(reference #306)

arikon committed:
Merge branch '0.5.x'

Conflicts:
	package.json

arikon committed:
Fix scripts to use q 0.8.10

arikon committed:
Update npm-shrinkwrap.json

Commit:
Fixed npmPackages check (close #300)

Commit:
Little code clean up

scf2k committed:
Merge pull request #298 from bem/inspector-expand

Arch inspector allows to expand and collapse nodes

arikon committed:
Merge branch '0.5.x'

arikon commented:
@narqo Актуализируешь?

arikon committed:
Merge pull request #301 from bem/documentation

Описал пример поключения локализационных технологий сборки

arikon committed:
Merge branch '0.5.x'

arikon committed:
Merge branch '0.5.x'

arikon committed:
Use istanbul as a test coverage tool (Close #287)

Commit:
Comment added

Commit:
Fixed typo in jsdoc

Commit:
SvnLibraryNode: Ability to map checkout paths to custom directories (reference #311)

arikon committed:
Update README.md and README.ru.md

arikon committed:
bem create: Support b__e_m form of declaration of BEM-entities to create

arikon committed:
Merge pull request #319 from bem/bem-create-ids

bem create: Support b__e_m form of declaration of BEM-entities to create

arikon committed:
css tech: Wrong selector for modifier without value in created file

arikon committed:
Merge pull request #321 from bem/css-mod-selector

css tech: Wrong selector for modifier without value in created file

arikon committed:
bem create: Allow techs specifying using dot notation (block__elem.css,js)

arikon committed:
Merge pull request #322 from bem/bem-create-techs

bem create: Allow techs specifying using dot notation

arikon committed:
Replace link to BEM methodology

arikon committed:
Depends on coa ~0.3.9

arikon committed:
Depend on q ~0.8.11

tadatuta committed:
docs: fix typo

arikon committed:
Replace Q.nbind() with Q.nfbind()

arikon committed:
Fix bug in util.bemType(); add tests for util.bemType()

arikon committed:
Fix util.bemParseKey() to work with multicomponent tech extension

arikon committed:
Refactor SvnLibraryNode to avoid copy-paste

arikon committed:
Merge pull request #311 from bem/i311

SvnLibraryNode: Ability to map checkout paths to custom directories

arikon committed:
A bit more cleaner formatting in BundleNode

arikon committed:
Merge pull request #309 from bem/merged-issue306

Magic nodes doesn't link the nodes it creates with parent magic nodes

arikon committed:
Broken link at bem.info for techs (bem/bem-www#36)

tadatuta committed:
ie6.css tech: Don't include bundle.css

Commit:
Refer correct npm version in the documentation.

Commit:
Backport 0.5.x into dev

Conflicts:
	npm-shrinkwrap.json
	package.json

eprev committed:
Fixed target for coverage

Commit:
Fix for bem server on windows (reference #211)

Conflicts:
	npm-shrinkwrap.json

Commit:
Merge branch 'i-87' into dev

Part of the default-arch is moved to standalone node (SeedNode) (close #87)
Libraries utilize the same build initialization as project - it's possible to build blocks levels within libraries (close #52)

Commit:
Merge branch 'merged-bundle' into dev

MergedBundle allows to override the list of the bundles to use.
The tech to merge (bemdecl.js, deps.js) can be overridden.

eprev committed:
Ability to specify source techs for BundlesLevelNode

eprev committed:
Sync with bem:custom-bundle-source

eprev committed:
Fixed function call

tadatuta committed:
Documentation: Update installation instructions

scf2k committed:
Merge pull request #351 from bem/docInstallationUpdate

Documentation: Update installation instructions

eprev committed:
jsdoc

scf2k committed:
Merge pull request #349 from eprev/source-techs

Add method to BundlesLevelNode to specify source techs for bundles

arikon committed:
Merge branch 'typo' into dev

arikon committed:
Merge branch 'bem-create-techs' into dev

veged committed:
Merge pull request #347 from eprev/patch-1

Fixed target for coverage

Commit:
deps.js: don't swallow parsing errors

arikon commented:
@narqo Отребейзи от `dev`, пожалуйста

narqo commented:
Наверное это надо закрыть и отправить pull-реквест в `dev` (8 месяцев назад девом был мастер)?

\cc @scf2k

scf2k commented:
Да, мастер стал девом.

scf2k commented:
Пул реквест был и можно закрыть или его так и не было?

narqo commented:
Вроде это в #424 сделали
